### PR TITLE
Infrastructure

### PR DESCRIPTION
**Pull request by jamieparkinson**
This makes the log forwarder work, and deploys it from CI - I'm using it as a self-contained example to experiment a bit with how we might deploy (non-Dockerised) lambdas.

### CODE REVIEW

File: .buildkite/scripts/autoformat.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+set -euo pipefail
```
Comment (alexwlchan) [inline .buildkite/scripts/autoformat.sh]: nit: for scripts I slightly prefer the longer names, more clear to folks not familiar with bash and easier to google:

```suggestion
set -o errexit
set -o nounset
set -o pipefail
```

File: .buildkite/scripts/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+TERRAFORM_ROLE="arn:aws:iam::760097843905:role/platform-ci"
+
+# Do this here rather than in Dockerfile so we don't install it on every build/test
+apt install -y awscli zip
```
Comment (alexwlchan) [inline .buildkite/scripts/deploy.sh:7]: Is this going to install on the host instances? Colour me a bit surprised they're not there already.

File: .buildkite/scripts/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+cd $ROOT/terraform
+terraform init -backend-config="role_arn=${TERRAFORM_ROLE}"
+S3_BUCKET=$(terraform output -raw s3_bucket)
+S3_KEY=$(terraform output -raw s3_key)
+FUNCTION_NAME=$(terraform output -raw function_name)
+cd $ROOT
+
+yarn package
+aws s3 cp $ROOT/package.zip "s3://${S3_BUCKET}/${S3_KEY}"
+VERSION_ID=$(aws s3api list-object-versions --bucket $S3_BUCKET --prefix $S3_KEY --query 'Versions[?IsLatest].[VersionId]' --output text)
```
Comment (alexwlchan) [inline .buildkite/scripts/deploy.sh]: I think you can do line breaks inside `$(…)`? Let's add some for readability:

```suggestion
VERSION_ID=$(aws s3api list-object-versions \
  --bucket $S3_BUCKET \
  --prefix $S3_KEY \
  --query 'Versions[?IsLatest].[VersionId]' \
  --output text)
```

File: src/handler.ts [hunk 4]
```diff
@@ -25,11 +26,26 @@ export const createHandler =
...
     });
+
+    if (failures.length !== 0) {
+      const failedServices = new Set(
+        failures.map(({ document }) => document.service)
+      );
+      console.error(
+        `Failed to ingest documents from ${failedServices.size} services: ${[
+          ...failedServices,
+        ].join(", ")}`
```
Comment (alexwlchan) [inline src/handler.ts:47]: Why do we need the spread operator here? I don't think I understand what the difference is between this and:

```suggestion
        `Failed to ingest documents from ${failedServices.size} services: ${failedServices.join(", ")}`
```

File: terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  s3_bucket         = aws_s3_object.log_forwarder_package.bucket
+  s3_key            = aws_s3_object.log_forwarder_package.key
+  s3_object_version = aws_s3_object.log_forwarder_package.version_id
+  handler           = "lambda.handler"
+
+  timeout                 = 15 * 60 // 15 minutes
+  memory_size             = 512
+  runtime                 = "nodejs18.x"
+  forward_logs_to_elastic = false // Prevent loops
```
Comment (alexwlchan) [inline terraform/main.tf:15]: ➿ ➿ ➿

File: terraform/terraform.tf [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    region         = "eu-west-1"
+  }
+}
+
+provider "aws" {
+  region = "eu-west-1"
+
+  default_tags {
+    tags = {
+      TerraformConfigurationURL = "https://github.com/wellcomecollection/elasticsearch-log-forwarder/tree/main/terraform"
```
Comment (alexwlchan) [inline terraform/terraform.tf:19]: 💯

**Review by alexwlchan — APPROVED** [review]

File: .buildkite/scripts/deploy.sh (hunk 2 above)
Comment (jamieparkinson) [reply .buildkite/scripts/deploy.sh:7]: Inside the docker container - basically had to choose between installing this in the container, or node (etc etc) on the host. Not ideal, but straightforward.

File: src/handler.ts (hunk 4 above)
Comment (jamieparkinson) [reply src/handler.ts:47]: Because it's a Set, and needs to be spread into an array to have access to array ops

**Pull request merged**